test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1

```diff
@@ -715,3 +715,209 @@ Describe 'Base type has abstract properties' -Tags "CI" {
+                protected internal virtual string VirtualProperty2 { get; set; } = "C1_VirtualProperty2";
+                protected internal virtual string VirtualMethod1() { return "C1_VirtualMethod1"; }
+                protected internal virtual string VirtualMethod2() { return "C1_VirtualMethod2"; }
+
+                public string CtorUsed {  get; set; }
+                public C1ProtectedInternal() { CtorUsed = "default ctor"; }
+                protected internal C1ProtectedInternal(string p1) { CtorUsed = "C1_ctor_1args:" + p1; }
+            }
+'@
+        $c2DefinitionProtectedInternal = @'
+            class C2ProtectedInternal : C1ProtectedInternal {
+                C2ProtectedInternal() : base() { $this.VirtualProperty2 = 'C2_VirtualProperty2' }
+                C2ProtectedInternal([string]$p1) : base($p1) { $this.VirtualProperty2 = 'C2_VirtualProperty2' }
+
+                [string]GetInstanceField() { return $this.InstanceField }
+                [string]SetInstanceField([string]$value) { $this.InstanceField = $value; return $this.InstanceField }
+                [string]GetInstanceProperty() { return $this.InstanceProperty }
+                [string]SetInstanceProperty([string]$value) { $this.InstanceProperty = $value; return $this.InstanceProperty }
+                [string]CallInstanceMethod() { return $this.InstanceMethod() }
+
+                [string]GetVirtualProperty1() { return $this.VirtualProperty1 }
+                [string]SetVirtualProperty1([string]$value) { $this.VirtualProperty1 = $value; return $this.VirtualProperty1 }
+                [string]CallVirtualMethod1() { return $this.VirtualMethod1() }
+
+                [string]$VirtualProperty2
+                [string]VirtualMethod2() { return 'C2_VirtualMethod2' }
+                # Note: Overriding a virtual property in a derived PowerShell class prevents access to the
+                #       base property via simple typecast ([base]$this).VirtualProperty2.
+                [string]GetVirtualProperty2() { return $this.VirtualProperty2 }
+                [string]SetVirtualProperty2([string]$value) { $this.VirtualProperty2 = $value; return $this.VirtualProperty2 }
+                [string]CallVirtualMethod2Base() { return ([C1ProtectedInternal]$this).VirtualMethod2() }
+                [string]CallVirtualMethod2Derived() { return $this.VirtualMethod2() }
+
+                [string]GetInstanceMemberDynamic([string]$name) { return $this.$name }
+                [string]SetInstanceMemberDynamic([string]$name, [string]$value) { $this.$name = $value; return $this.$name }
+                [string]CallInstanceMemberDynamic([string]$name) { return $this.$name() }
+            }
+
+            [C2ProtectedInternal]
+'@
+
+        Add-Type -TypeDefinition $c1DefinitionProtectedInternal
+        Add-Type -TypeDefinition (($c1DefinitionProtectedInternal -creplace 'C1ProtectedInternal', 'C1Protected') -creplace 'protected internal', 'protected')
```

Comment:
Let's run `Add-Type` only once per process, so as to avoid error when running the tests again.
```suggestion
        if (-not ('C1ProtectedInternal' -as [type])) {
            Add-Type -TypeDefinition $c1DefinitionProtectedInternal
            Add-Type -TypeDefinition (($c1DefinitionProtectedInternal -creplace 'C1ProtectedInternal', 'C1Protected') -creplace 'protected internal', 'protected')
        }
```

Reply:
Adding the exact same type definition again doesn't produce an error (the definition is hashed if I recall correctly). And if I have made changes to the definition, I don't want a re-run of the tests with the old definition.

Reply:
You are right, I missed that :)